### PR TITLE
[Feature] Add module test  process of mobilenetv3

### PR DESCRIPTION
**Pull request by shiyutang**

### CODE REVIEW

File: tutorials/mobilenetv3_prod/README.md [hunk 1]
```diff
@@ -1,36 +1,440 @@
...
+    - [4.2 数据加载对齐]()
+    - [4.3 评估指标对齐]()
+    - [4.4 损失对齐]()
+    - [4.5 反向梯度初次对齐]()
+    - [4.6 训练对齐]()
+
+
+## 1. 简介
+
+本部分内容包含基于 [MobileNetV3](https://arxiv.org/abs/1905.02244) 的复现对齐过程，可以结合[论文复现指南]()进行学习。
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 这里补充下具体的目录吧

File: tutorials/mobilenetv3_prod/README.md [hunk 2]
```diff
@@ -1,36 +1,440 @@
...
+
+## 2. 复现流程
+在论文复现中我们可以根据网络训练的流程，将对齐流程划分为数据加载对齐、模型前向对齐、评估指标对齐、反向梯度对齐和训练对齐。其中不同对齐部分我们会在下方详细介绍。
+
+在对齐验证的流程中，我们依靠 reprod logger 日志工具查看 paddle 和官方同样输入下的输出是否相同，这样的查看方式具有标准统一，比较过程方便等优势。
+
+### 2.1 reprod logger 简介
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 统一修改为reprod_log吧，和repo名称保持一致

File: tutorials/mobilenetv3_prod/README.md [hunk 3]
```diff
@@ -1,36 +1,440 @@
...
+Reprod logger 是一个用于 numpy 数据记录和对比工具，通过传入需要对比的两个 numpy 数组就可以在指定的规则下得到数据之差是否满足期望的结论。其主要接口的说明可以看它的 [github 主页](https://github.com/WenmuZhou/reprod_log)。
+
+## 3. 准备数据和环境
+在进行我们的对齐验证之前，我们需要准备运行环境、用于输入的伪数据、paddle 模型参数和官方模型权重参数。
+
+### 3.1 准备环境
+* 克隆本项目
+
+```bash
+https://github.com/PaddlePaddle/models.git
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: git clone

File: tutorials/mobilenetv3_prod/README.md [hunk 4]
```diff
@@ -1,36 +1,440 @@
...
+cd model/tutorials/mobilenetv3_prod/
+```
+
+* 安装paddlepaddle [Paddle安装指南](https://www.paddlepaddle.org.cn/)
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: [Paddle安装指南](https://www.paddlepaddle.org.cn/)

这个内容放在pip命令的后面，如下：

更多版本或者环境下的安装可以参考：[Paddle安装指南](https://www.paddlepaddle.org.cn/)

File: tutorials/mobilenetv3_prod/README.md [hunk 5]
```diff
@@ -1,36 +1,440 @@
...
+    diff_helper = ReprodDiffHelper()
+    torch_info = diff_helper.load_info("./result/forward_torch.npy")
+    paddle_info = diff_helper.load_info("./result/forward_paddle.npy")
+
+    # compare result and produce log
+    diff_helper.compare_info(torch_info, paddle_info)
+    diff_helper.report(path="./result/log/forward_diff.log")
+```
+
+在代码示例中也可以学习到 reprod logger的主要接口，包含add、save、load_infor、compare_infor、report的用法。
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: reprod logger -> reprod_log

File: tutorials/mobilenetv3_prod/README.md [hunk 6]
```diff
@@ -1,36 +1,440 @@
...
+            inputs, dtype=torch.float32),
+        torch.tensor(
+            labels, dtype=torch.int64),
+        torch_model,
+        accuracy_torch,
+        'ref', reprod_logger)
+```
+这部分模型和输入的导入均和之前一致，我们增加了模型计算评估指标的部分，由于我们之前验证了模型的输入一致。那么也就是评估指标的输入相同，我们只需要对比输出是否一致，即可确定评估指标的实现是否正确。
+
+【**运行文件**】
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 这种期望换行的格式，在md文件中都是需要空行的，需要统一修改下

File: tutorials/mobilenetv3_prod/README.md [hunk 7]
```diff
@@ -1,36 +1,440 @@
...
+通过运行以下代码，我们验证评估指标对齐效果。
+```bash
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 代码片段前空一行把

File: tutorials/mobilenetv3_prod/README.md [hunk 8]
```diff
@@ -1,36 +1,440 @@
...
+【**获得结果**】
+
+进入`result/log/loss_diff.log`中，就会有下列结果，而结果说明我们评估指标的实现正确，完成第三个打卡点：
+```bash
+[2021/12/22 20:13:41] root INFO: loss:
+[2021/12/22 20:13:41] root INFO:     mean diff: check passed: True, value: 0.0
+[2021/12/22 20:13:41] root INFO: diff check passed
+```
+
+### 4.4 反向梯度对齐
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 4.5

File: tutorials/mobilenetv3_prod/README.md [hunk 9]
```diff
@@ -1,36 +1,440 @@
...
+结合模型和损失，我们就可以验证反向过程，反向梯度传导的是否正确包含了优化器，学习率以及梯度的计算，而验证过程只需要多观察几轮损失即可明确反向是否正确传导，主要验证流程如下所示：
+<div align="center">
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 图片内容之前也都空一行

File: tutorials/mobilenetv3_prod/README.md [hunk 10]
```diff
@@ -1,36 +1,440 @@
...
+    <img src="./images/backward.png" width=500">
+</div>
+
+以上参考流程可以使用以下代码实现：
+```python
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/README.md]: 代码片段前空一行，该处内容需要统一修改下

File: tutorials/mobilenetv3_prod/mobilenetv3_ref/train.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+        '--apex-opt-level',
+        default='O1',
+        type=str,
+        help='For apex mixed precision training'
+        'O0 for FP32 training, O1 for mixed precision training.'
+        'For further detail, see https://github.com/NVIDIA/apex/tree/master/examples/imagenet'
+    )
+
+    # distributed training parameters
+    parser.add_argument(
```
Comment (littletomatodonkey) [inline tutorials/mobilenetv3_prod/mobilenetv3_ref/train.py]: 应该需要加个参数，不然多卡启动会报错？

```
parser.add_argument("--local_rank", type=int, default=0)
```

File: tutorials/mobilenetv3_prod/README.md (hunk 1 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: 论文复现指南还没有合入，可以这边先合入了再加，因为复现指南中也需要加这边的link

File: tutorials/mobilenetv3_prod/README.md (hunk 2 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 3 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 4 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 5 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 6 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 7 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: 这个好像对markdown结果没有影响，但是还是空了~

File: tutorials/mobilenetv3_prod/README.md (hunk 8 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 9 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/README.md (hunk 10 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/README.md]: done

File: tutorials/mobilenetv3_prod/mobilenetv3_ref/train.py (hunk 11 above)
Comment (shiyutang) [reply tutorials/mobilenetv3_prod/mobilenetv3_ref/train.py]: done

**Review by littletomatodonkey — APPROVED** [review]
